### PR TITLE
[feat] Support model parallelism in OSS

### PR DESCRIPTION
**Pull request by joshim5**
This PR adds support for model parallelism in OSS. I added an [example integration with Fairseq](https://github.com/fairinternal/fairseq-py/pull/1531/files#diff-6c1200391a37100c04391c1bbee841eaa87808c15fb74bde964e0bccfde668c6R77).

The key idea is to expose `filter_params_fn` which can filter out weights that are replicated on various model parallel units. See below for an example of how this can be used. I personally think `_filter_model_parallel_params_fn` should live inside the user codebase (e.g. Fairseq) since it makes assumptions about the model architecture, but I'm also happy to add it here. The megatron code is used by many downstream codebases (including Fairscale) and should therefore appropriately mark weights with a `model_parallel` attribute.

```
        def _filter_model_parallel_params_fn(params):
            mp_rank_is_zero = (get_model_parallel_rank() == 0)
            params = list(params)
            parameters_for_norm = list(filter(
                lambda p: getattr(p, 'model_parallel', False) or mp_rank_is_zero, params))
            return parameters_for_norm
 ```

### CODE REVIEW

**Comment by haven-jeon** [comment]
@joshim5 can I see `example integration with Fairseq`. I can't access it.

**Comment by joshim5** [comment]
@haven-jeon: the Fairseq integration hasn't been merged to our public repo yet. Anything we can help with in the meantime?

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -252,11 +254,19 @@ def clip_grad_norm(self, max_norm: Union[float, int], norm_type: Union[float, in
...
+        if filter_params_fn is not None:
+            local_params = filter_params_fn(local_params)
+
         # Compute the norm on this grad set,
         # then sync all the norms from all ranks
         if norm_type == inf:
             total_norm = max(p.grad.detach().abs().max().to(self._device) for p in local_params)  # type: ignore
-            dist.all_reduce(total_norm, op=torch.distributed.ReduceOp.MAX, group=self.group)
+            # all reduce over data parallel and model parallel workers
+            dist.all_reduce(total_norm, op=torch.distributed.ReduceOp.MAX)
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:269]: would it make sense to explicity pass "world" to make it clear ?

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -206,6 +206,8 @@ def cleanup() -> None:
                 # Opportunistically try to empty the queue
                 optimizer._try_consume_work_handle()
 
+                logging.warning("reducing one parameter")
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:209]: my bad, I can remove that

**Review by blefaudeux — APPROVED** [review]
LGTM, I can do the minor nits if you'd like, thanks a lot @joshim5 for this work ! We can also wait for the fairseq unit test to be cleared if you prefer, fine by me

File: fairscale/optim/oss.py (hunk 1 above)
Comment (joshim5) [reply fairscale/optim/oss.py:269]:
```suggestion
            dist.all_reduce(total_norm, op=torch.distributed.ReduceOp.MAX, group=dist.group.WORLD)
```
Sure!

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 2 above)
Comment (joshim5) [reply fairscale/nn/data_parallel/sharded_ddp.py:209]:
```suggestion
```

**Comment by joshim5** [comment]
The failing Fairseq unit test is about checkpoint saving/reloading, so it's orthogonal to this PR, which adds support for model parallelism with OSS.

Here's a demonstration in Fairseq ([using my linked PR above](https://github.com/fairinternal/fairseq-py/pull/1531/files#diff-6c1200391a37100c04391c1bbee841eaa87808c15fb74bde964e0bccfde668c6R77)) showing that model parallelism gives the same result using:
- sharded ddp, optimizer state sharding (ddp sharded, os)
- c10d ddp, optimizer state sharding (ddp c10d, os)
- c10d ddp, no state sharding (ddp c10d, none)

Note that loss and gnorm are the same in all runs.

Since the test passes, I'm comfortable merging this PR.

```
## Model parallelism
Tiny model distributed over two GPUs. Experiments were doen on 8 V100s, single node, memory efficient fp16.

### ddp sharded, os
python train.py --task masked_lm /checkpoint/bioseq_nonsecure/model-parallel-data/tiny_sample_valid_ur50-bin  --dataset-impl fasta  --save-dir checkpoints/zero-0-mp-1    --dropout 0.1   --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0   --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07   --tokens-per-sample 128 --sample-break-mode none   --max-tokens 128 --memory-efficient-fp16 --no-progress-bar --log-interval 20 --seed 4 --max-epoch 1 --max-update 50 --encoder-layers 4 --no-save  --arch model_parallel_roberta_large --model-parallel-size 2 --update-freq 2 --zero-sharding os

2020-12-22 07:11:17 | INFO | train_inner | epoch 001:     20 / 78 loss=0.979, ppl=1.97, wps=16454, ups=16.06, wpb=1024, bsz=8, num_updates=20, lr=2.5995e-06, gnorm=2.667, loss_scale=128, train_wall=3, wall=10
2020-12-22 07:11:18 | INFO | train_inner | epoch 001:     40 / 78 loss=0.882, ppl=1.84, wps=17198.4, ups=16.9, wpb=1017.7, bsz=8, num_updates=40, lr=5.099e-06, gnorm=2.146, loss_scale=128, train_wall=1, wall=11
2020-12-22 07:11:19 | INFO | fairseq_cli.train | begin validation on "valid" subset
2020-12-22 07:11:27 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 0.708 | ppl 1.63 | wps 30793.4 | wpb 509.6 | bsz 4 | num_updates 50
2020-12-22 07:11:27 | INFO | fairseq_cli.train | begin save checkpoint
2020-12-22 07:11:27 | INFO | fairseq_cli.train | end of epoch 1 (average epoch stats below)
2020-12-22 07:11:27 | INFO | train | epoch 001 | loss 0.896 | ppl 1.86 | wps 4436.5 | ups 4.34 | wpb 1021.5 | bsz 8 | num_updates 50 | lr 6.34875e-06 | gnorm 2.239 | loss_scale 128 | train_wall 4 | wall 20
2020-12-22 07:11:27 | INFO | fairseq_cli.train | done training in 19.0 seconds


### ddp c10d, os

python train.py --task masked_lm /checkpoint/bioseq_nonsecure/model-parallel-data/tiny_sample_valid_ur50-bin  --dataset-impl fasta  --save-dir checkpoints/zero-0-mp-1    --dropout 0.1   --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0   --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07   --tokens-per-sample 128 --sample-break-mode none   --max-tokens 128 --memory-efficient-fp16 --no-progress-bar --log-interval 20 --seed 4 --max-epoch 1 --max-update 50 --encoder-layers 4 --no-save  --arch model_parallel_roberta_large --model-parallel-size 2 --update-freq 2 --zero-sharding os

2020-12-21 19:10:03 | INFO | train_inner | epoch 001:     20 / 78 loss=0.979, ppl=1.97, wps=17043.9, ups=16.64, wpb=1024, bsz=8, num_updates=20, lr=2.5995e-06, gnorm=2.667, loss_scale=128, train_wall=3, wall=9
2020-12-21 19:10:05 | INFO | train_inner | epoch 001:     40 / 78 loss=0.882, ppl=1.84, wps=17360.1, ups=17.06, wpb=1017.7, bsz=8, num_updates=40, lr=5.099e-06, gnorm=2.146, loss_scale=128, train_wall=1, wall=10
2020-12-21 19:10:05 | INFO | fairseq_cli.train | begin validation on "valid" subset
2020-12-21 19:10:14 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 0.708 | ppl 1.63 | wps 32323.8 | wpb 509.6 | bsz 4 | num_updates 50
2020-12-21 19:10:14 | INFO | fairseq_cli.train | begin save checkpoint
2020-12-21 19:10:14 | INFO | fairseq_cli.train | end of epoch 1 (average epoch stats below)
2020-12-21 19:10:14 | INFO | train | epoch 001 | loss 0.896 | ppl 1.86 | wps 4358.1 | ups 4.27 | wpb 1021.5 | bsz 8 | num_updates 50 | lr 6.34875e-06 | gnorm 2.239 | loss_scale 128 | train_wall 4 | wall 20
2020-12-21 19:10:14 | INFO | fairseq_cli.train | done training in 18.8 seconds


### ddp c10d, none

python train.py --task masked_lm /checkpoint/bioseq_nonsecure/model-parallel-data/tiny_sample_valid_ur50-bin  --dataset-impl fasta  --save-dir checkpoints/zero-0-mp-1    --dropout 0.1   --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0   --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07   --tokens-per-sample 128 --sample-break-mode none   --max-tokens 128 --memory-efficient-fp16 --no-progress-bar --log-interval 20 --seed 4 --max-epoch 1 --max-update 50 --encoder-layers 4 --no-save  --arch model_parallel_roberta_large --model-parallel-size 2 --update-freq 2

2020-12-21 12:46:14 | INFO | train_inner | epoch 001:     20 / 78 loss=0.979, ppl=1.97, wps=3425.2, ups=3.34, wpb=1024, bsz=8, num_updates=20, lr=2.5995e-06, gnorm=2.667, loss_scale=128, train_wall=8, wall=15
2020-12-21 12:46:20 | INFO | train_inner | epoch 001:     40 / 78 loss=0.882, ppl=1.84, wps=3369, ups=3.31, wpb=1017.7, bsz=8, num_updates=40, lr=5.099e-06, gnorm=2.146, loss_scale=128, train_wall=6, wall=21
2020-12-21 12:46:23 | INFO | fairseq_cli.train | begin validation on "valid" subset
2020-12-21 12:46:39 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 0.708 | ppl 1.63 | wps 8253.6 | wpb 509.6 | bsz 4 | num_updates 50
2020-12-21 12:46:39 | INFO | fairseq_cli.train | begin save checkpoint
2020-12-21 12:46:39 | INFO | fairseq_cli.train | end of epoch 1 (average epoch stats below)
2020-12-21 12:46:39 | INFO | train | epoch 001 | loss 0.896 | ppl 1.86 | wps 1636.9 | ups 1.6 | wpb 1021.5 | bsz 8 | num_updates 50 | lr 6.34875e-06 | gnorm 2.239 | loss_scale 128 | train_wall 17 | wall 40
2020-12-21 12:46:39 | INFO | fairseq_cli.train | done training in 39.2 seconds

```

**Comment by blefaudeux** [comment]
Looks great to me, thanks again @joshim5 !
